Ivy/di/inherited services

alxhub opened the pull request.

alxhub committed:
refactor(ivy): fix type of factory functions to allow explicit types

Factory functions written by the compiler optionally allow an explicit type
to be passed. If called with this type, an instance of the given type will
be created instead of the type for which the factory was generated. This is
used to power inheritance of Angular types, as if the constructor of a class
is inherited from its superclass, then the factory function of the
superclass must be used (it has all the DI info) to construct an instance of
the derived class.

This commit adjusts typings in a few places to allow factory functions to be
called with this extra type parameter.

alxhub committed:
fix(ivy): call factory functions with correct type for derived classes

In a derived service class with no decorator (and therefore no factory) of
its own, the factory function of the base class will be used instead.
Previously this logic had a bug where the factory function would be called
with no arguments, which would incorrectly create an instance of the base
class.

This commit adds logic to call the base class' factory and pass the type of
the derived class, which will correctly construct an instance of the
derived class using the base class' factory. A test is also added to verify
correctness of this behavior.

gkalpak commented on packages/core/src/render3/di.ts:
```diff
@@ -637,8 +637,8 @@ export class NodeInjector implements Injector {
-export function ɵɵgetFactoryOf<T>(type: Type<any>): ((type: Type<T>| null) => T)|null {
-  const typeAny = type as any;
+export function ɵɵgetFactoryOf<T>(type: Type<any>): FactoryFn<T>|null {
+  const typeAny = `type ` as any;
```
Is this intentional? It looks weird :grin:

alxhub replied:
Definitely not intentional, haha. Thanks for catching!

alxhub committed:
feat(ivy): require 'token' for ngInjectableDef

The compiler generates a 'token' field when it emits an ngInjectableDef,
but this field was not required by defineInjectable or the InjectableDef
interface, nor was it added by InjectionToken.

This commit makes 'token' required and adds it where missing.

alxhub committed:
fix(ivy): distinguish inherited injectable defs by token

In older browsers such as IE10, inheritance of static properties on class
types is implemented by copying the properties. This makes hasOwnProperty
unreliable for checking if a static property is defined directly on a class
or is inherited. This causes problems when trying to read the
ngInjectableDef properties of inherited services, as in IE10 even inherited
definitions will appear to be local.

Instead, the 'token' property on ngInjectableDef can be leveraged to detect
when a definition is read which doesn't belong to the type itself.

alxhub commented:
[Presubmit](https://test.corp.google.com/ui#id=OCL:251511177:BASE:252117320:1559942216770:9be37a72)

[Ivy Presubmit](http://test/OCL:251511177:BASE:252136514:1559948338752:95d3cc16)

kara commented on packages/core/src/di/interface/defs.ts:
```diff
@@ -181,7 +190,15 @@ export function ɵɵdefineInjector(options: {factory: () => any, providers?: any
+  //
+  // Instead, the ngInjectableDef's token is compared to the type, and if they don't match then the
+  // property was not defined directly on the type itself, and was likely inherited. The definition
+  // is only returned if the type matches the def.token.
```
Thanks for the comment 👍

kara reviewed (approved):
LGTM
